### DISCUSSION TITLE
Autonomous Control not working, /dev/null is an unused topic?

### DISCUSSION THREAD

**Discussion by KCHJ1999** [prl-mushr/mushr#70] [Software]
Hi, I am a student doing my final year project this semester and my project topic is to replicate a MuSHR car but I came across some errors on the autonomous control.

I followed the instructions in the tutorial for the Autonomous Navigation, cloned the RHC and PF and everything is working fine but when i tried to run RVIZ to test the autonomous controls, it does not tell me goal set after i clicked the 2D Nav Goal and just prompt me setting pose and setting goal.

Picture:
![WhatsApp Image 2022-01-19 at 10 58 48 AM (2)](https://user-images.githubusercontent.com/96906767/150056457-b08c7c17-47ff-4273-9ab7-e99b4d76034e.jpeg)

When i tried to check for the errors, i went to check the jot_teleop.yaml file and realised the codes for autonomous control topic name is /dev/nulll which is a unused topic, am I suppose to write any codes on it? Is that why the autonomous controls isnt working because there are no codes on it? Please help. Thank you.

Picture:
![WhatsApp Image 2022-01-19 at 10 58 48 AM](https://user-images.githubusercontent.com/96906767/150056347-5361bf84-81cb-4b7b-a6f7-7413ff834dc2.jpeg)

My RHC Controller prompt me this error, Am i suppose to change the map name to my current map name?

![WhatsApp Image 2022-01-19 at 10 58 48 AM (1)](https://user-images.githubusercontent.com/96906767/150056447-2788acfc-e108-4dc9-90e7-04c4b650fd7a.jpeg)

I'll be glad if someone could advise me on this. Thank you.

**Comment by schmittlema** [comment]
Hi @KCHJ1999. I think the RHC controller is actually doing some initial processing to generate a trajectory library. If you reduce K [here](https://github.com/prl-mushr/mushr_rhc/blob/master/mushr_rhc_ros/launch/params/trajgen/dispersion.yaml#L2) to say 30 or 60 it should take less time. Also a print statement in the for loop [here](https://github.com/prl-mushr/mushr_rhc/blob/master/mushr_rhc_ros/src/librhc/trajgen/dispersion.py#L129) could show you the progress. Let me know if that helps, or if something else is going astray!

**Reply by KCHJ1999** [reply]
Alright. Thank you for your advice. I will try it and get back to you.

**Reply by KCHJ1999** [reply]
Hello, i tried out your suggestions by changing the K to 30 and it still didn't register the autonomous controls. Is it normal for the car to vibrate in the simulator while the rhc and pf node are active? Thank you.

**Reply by schmittlema** [reply]
Vibration is normal, it is the particle filter deciding where the robot is.

Did you try adding a print statement? Did anything change in your terminal output?

**Reply by KCHJ1999** [reply]
Oh i see. Erm sorry i didnt add a print statement because i thought it was already inside. I will add it now and let you know the result.

I have another issue with the ilwifi firmware. I loaded up the image on the website and installed a Intel 8265NGW Wi-Fi card but i get this error message every time the jetson boots up. Do i need to install or update the firmware?
![WhatsApp Image 2022-01-26 at 9 56 14 AM](https://user-images.githubusercontent.com/96906767/151097989-6d987a5a-db7b-41c0-809a-f72f8a3634d6.jpeg)

![WhatsApp Image 2022-01-26 at 9 55 29 AM](https://user-images.githubusercontent.com/96906767/151097993-32538c7a-d7b8-4f53-b5a6-a6982a26a066.jpeg)

Could this be the reason why the autonomous connection isn't working?  whenever I SSH and use tmux my connection cuts off sometimes.

**Comment by KCHJ1999** [comment]
Hi, i did a for loop print statement and here are the results
![WhatsApp Image 2022-01-26 at 11 28 52 AM](https://user-images.githubusercontent.com/96906767/151101359-88b70e47-0de6-4af6-8c8c-2d1340d6f1a7.jpeg)
Is this the result you want?

**Reply by schmittlema** [reply]
Yes thank you, it looks like it is not making it the required number of trajectories. If you could try increasing the samples to 11 this may rectify the problem. If 11 doesn't work try 15.

In regards to the wifi issue I haven't seen that before. But googling the issue I found [this](https://www.cyberciti.biz/faq/intel-wifi-on-debian-linux-when-you-get-firmware-failed-to-load-iwlwifi-8265-36-error/) which looks like it would solve your problem?

**Reply by KCHJ1999** [reply]
Samples? do i reduce the K to 12? because k-1 is 11. Thank you

Yea. i am trying to solve the wifi issue. Thank you. Im currently looking at that website too.

**Reply by schmittlema** [reply]
Shoot I forgot to link where the sample param is: [here ](https://github.com/prl-mushr/mushr_rhc/blob/master/mushr_rhc_ros/launch/params/trajgen/dispersion.yaml#L9)

**Reply by KCHJ1999** [reply]
Thank you. I will change it and get back to you.

**Reply by KCHJ1999** [reply]
Hello, may i ask how do i backup the current image that i have? I used AOMEI partition assistant to clone the sd card image to another sd card but after flashing it, my jetson didn't boot the image up? May i know what software did you used to make a backup of your current image? I think my MBR is corrupted and the jetson does not know where to look for the boot sector. Thank you

**Reply by schmittlema** [reply]
There is not a proper way to do this. We are moving away from the mushr image because of this amongst other reasons. Regardless @planc509 may have the latest instructions to do this?

**Reply by palanc** [reply]
Hi @KCHJ1999 - I have had success using this guide: [https://www.jetsonhacks.com/2020/08/08/clone-sd-card-jetson-nano-and-xavier-nx/](https://www.jetsonhacks.com/2020/08/08/clone-sd-card-jetson-nano-and-xavier-nx/)

Let us know if that helps!

**Comment by KCHJ1999** [comment]
Hi could this be the reason why the autonomous controls isn't working? I had this warning messages once i start teleoperating the car.
![WhatsApp Image 2022-02-04 at 10 31 29 PM (3)](https://user-images.githubusercontent.com/96906767/152552318-79796a9f-6b01-4443-ba95-b3ba218c9aa5.jpeg)

![WhatsApp Image 2022-02-04 at 10 31 29 PM (2)](https://user-images.githubusercontent.com/96906767/152552325-1cfa449b-d3d7-40de-895c-106cd0bf4f85.jpeg)

![WhatsApp Image 2022-02-04 at 10 31 29 PM (1)](https://user-images.githubusercontent.com/96906767/152552332-d3c50d64-d162-4be3-96c8-cceba3c5b92f.jpeg)

**Reply by schmittlema** [reply]
These errors are unrelated to autonomous navigation. Mainly they are saying that your joystick is disconnected. Sometimes it takes a second for it to be detected. Steps to check things are working
- run teleop.launch
- run mushr_pf real.launch
- run mushr_rhc real.launch
- run rviz. check the robot's position is correct per the map in rviz. If not, set a "2D pose estimate" to where the robot is.
- Set a "2D nav goal". Subscribe to mushr_rhc/traj_choosen topic : "Add" --> "By topic" --> car/mushr_rhc/traj_choosen (I think). This should show a trajectory of where the robot wants to go. 
- hold down Rb on the controller and it should go.

Somewhere in those steps (presumably the last 2) is where you are experiencing issues. To help debug the mushr_rhc code output from that terminal (the one you ran roslaunch mushr_rhc real.launch) will be most useful for me.

**Reply by KCHJ1999** [reply]
Alright, thank you for the breakdown. i will try it and get back to you.

**Reply by KCHJ1999** [reply]
(When the car is on)
![WhatsApp Image 2022-02-06 at 5 07 46 PM](https://user-images.githubusercontent.com/96906767/152674375-fefddc3f-c7db-434f-9b69-b403e4b06bfc.jpeg)
Is this temperature ok for the system? do i need to install any fans onto the jetson to cool down the temperature and may i know what temperature range value is bad for the system when the car is on?

**Reply by schmittlema** [reply]
Would you mind posting this in a separate post? Just so we can keep this thread about the autonomous navigation stack. Thanks!

**Reply by KCHJ1999** [reply]
Alright

**Comment by KCHJ1999** [comment]
Is this the output you want for the rhc? I sent it through a text file
[rhcoutput.txt](https://github.com/prl-mushr/mushr/files/8020586/rhcoutput.txt)

**Comment by KCHJ1999** [comment]
This is my terminal output from my RVIZ and PF. I changed the params to 11 and 15 but it still remains the same.
![WhatsApp Image 2022-02-10 at 11 15 45 AM](https://user-images.githubusercontent.com/96906767/153331372-f35d1388-075a-4a5c-9d6b-9946c88862e4.jpeg)

**Reply by schmittlema** [reply]
Did you delete [this print statement?](https://github.com/prl-mushr/mushr/discussions/70#discussioncomment-2047928) Or is that not printing now?

**Reply by KCHJ1999** [reply]
I didnt delete, its still printing 29

**Reply by schmittlema** [reply]
To be honest, this mpc is buggy (if it not already apparent) and we will be focused on fixing it once we complete our switch to python 3.
I'm attaching another controller package you can use. This is much more interpretable and likely will work with a few modifications.

1. This goes in your catkin_ws/src. Then you should run `catkin_make`
2. To run it you will run `roslaunch ta_lab2 mpc_controller.launch`
3. It won't work by default. You will need to tell it to listen to the particle filter's inferred pose on line 88 in controllernode.py
4. You will also need to change line 95 to be the `/car/mux/ackermann_cmd_mux/input/navigation` topic. To make sure you have the right topic name use `rostopic list` and find the topic that matches. 
5. It currently accepts a path to follow (line 84, 85, and cb_path) but you want to just give it a goal point. You will need to change the subscriber to the `move_base_simple/goal` topic rviz outputs. This outputs a [PoseStamped](https://gitlab.cs.washington.edu/cse490r/19sp/ta_lab2/-/blob/master/src/controlnode.py#L84) datatype. If you run into issues with getting the messages see this [pub/sub python tutorial](http://wiki.ros.org/ROS/Tutorials/WritingPublisherSubscriber%28python%29).
6. Once you can get a goal message you can set the path to be just one point and (hopefully) it tries to follow that point.

The files you should mainly be looking at are `mpc_controller.launch`, `controller.py`, `main.py`, `controllernode.py`. All mentioned lines are in `controllernode.py`

[ta_lab2-master.zip](https://github.com/prl-mushr/mushr/files/8042202/ta_lab2-master.zip)

**Reply by KCHJ1999** [reply]
Oh, I see. I  was worried because I initially thought it was the codes that causes it to not operate. Thank you for clearing my doubts. May I know when MuSHR migrates to Python 3? I will do the amendments and inform you if there are any errors. May i know which ros wiki do I go to understand your codes?

May I ask if there is anyone having similar problems relating to my issues currently from your side? Thank you.

**Reply by schmittlema** [reply]
1. Very soon hopefully. We are trying to switch our visualizer to foxglove and add Mac support.
2. This code does not have a ros wiki. It is fairly readable though
3. We have had similar problems yes. But unfortunately, we are short staffed and fixing the controller is lower priority to switching to python3

**Reply by KCHJ1999** [reply]
Do you think its the code that has the issue? why is it buggy?

**Reply by KCHJ1999** [reply]
i tried the above packages that you sent.It still didn't get set a path to the goal.

**Reply by schmittlema** [reply]
Yes the code is buggy likely because it has not been actively maintained for some time. In regards to the package I sent, did you make the changes I suggested to the codebase?

**Reply by KCHJ1999** [reply]
Yes i did make the changes to the code but, it still did not plan a trajectory or a goal.

**Reply by schmittlema** [reply]
I'm not sure what exactly is going wrong, but we have a student working on modifying that code to work. In the meantime, I would recommend you look at the slides on Model Predictive Control (Feb 12, Feb 14) to understand the concepts then try to debug what is going on. https://courses.cs.washington.edu/courses/cse478/20wi/
